### PR TITLE
Enable write flow for FDW for non-text/csv formats

### PR DESCRIPTION
**Pull request by denalex**
This PR introduces support for write flow for FDW for formats that are not `text` or `csv` (profiles are not `*:text` or `*:csv`). Previously, records in TEXT/CSV formats were processed using a pass-through input stream, where no deserialization took place and bytes were just copied to an external system. Records non in TEXT/CSV format were sent using `gpdbwritable` formatter / wire serialization protocol, which was not and is still not supported with FDW.

In FDW data from Greenplum is sent using CSV format, no matter what destination format it will end up in (e.g parquet or avro). This PR introduces a CSV deserializer to allow for parsing of CSV-formatted data arriving from Greenplum. It introduces a concept of a `RecordReader` and provides 3 implementations for Text, GPDBWritable and InputStream.

The PR also:
- introduces `GPDataGenerator` utility to generate a sample dataset and a SQL script for processing it. The new unit test depends on it and works with the dataset that has been generated using this utility.
- adds handling of new to GP7 FDW APIs so that a user can COPY TO <foreign table> backed by PXF.
- fixes a problem where segments that had no data were still creating creating empty files in external system

Finally, the PR enables a few integration tests for write data flow to exercise the new functionality.

### CODE REVIEW

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/bridge/WriteBridge.java [hunk 1]
```diff
@@ -76,14 +94,13 @@ public boolean setNext(DataInputStream inputStream) throws Exception {
         if (onerow == null) {
             return false;
         }
-        if (!accessor.writeNextObject(onerow)) {
-            throw new BadRecordException();
-        }
-        return true;
+        // TODO: can we do this ?
```
Comment (ashuka24) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/bridge/WriteBridge.java:97]: if this todo is still valid, can we be a little more descriptive about what exactly is the ask?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/RecordReaderFactory.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            case GPDBWritable:
+                return new GPDBWritableRecordReader(context);
+            case TEXT:
+                if (canHandleInputStream) {
+                    /*
+                    If downstream components (resolver / accessor) can handle an inputStream directly, use a shortcut
+                    to avoid reading bytes from the inputStream here and instead pass the inputStream in the record.
+                    This code used to use the Text class to read bytes until a line delimiter was found. This would cause
+                    issues with wide rows that had 1MB+, because the Text code grows the array to fit data, and
+                    it does it inefficiently. We observed multiple calls to System.arraycopy in the setCapacity method
```
Comment (ashuka24) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/RecordReaderFactory.java:48]: nit: was hard to follow which `it` referred to what
```suggestion
                    it does so inefficiently. We observed multiple calls to System.arraycopy in the setCapacity method

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    public List<OneField> readRecord(DataInput input) throws Exception {
+        if (iterator == null) {
+            iterator = parser.iterateRecords((InputStream) input, context.getDatabaseEncoding()).iterator();
+        }
+
+        if (!iterator.hasNext()) {
+            return null; // no more data to read
+        }
+
+        // this can only be done after iterator read a record or called hasNext()
```
Comment (ashuka24) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:127]: can you explain why this can only be done after the iterator has read a record? if we have not initialized anything, how can we call `hasNext`?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        return record;
+    }
+
+    private Class getJavaClass(DataType dataType) {
+        // only very specific numeric types will get their own functions
+        // all other data types are considered as Strings
+        switch (dataType) {
+            case BOOLEAN:
+                return Boolean.class;
+//            case BYTEA:
```
Comment (ashuka24) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:165]: why is this commented out?

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        }
+    }
+
+    /**
+     * A POJO to capture all metadata about a column of a dataset.
+     */
+    @Data
+    @AllArgsConstructor
+    private static class Column {
+        String name;               // name     of the column            , used to create SQL DDL
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java:91]:
```suggestion
        String name;               // name of the column                , used to create SQL DDL

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                    assertNull(field.val, String.format("Expected null in row=%d col=%d", row, col));
+                } else {
+                    if (COLUMNS[col].getClazz() == ByteBuffer.class) {
+                        assertTrue(ByteBuffer.class.isAssignableFrom(field.val.getClass()));
+                    } else {
+                        assertSame(COLUMNS[col].getClazz(), field.val.getClass(), String.format("Java class mismatch row=%d col=%d", row, col));
+                    }
+                    // make adjustments to expected values
+                    Object expected = cells.get(col);
+                    if (col == 25) {
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java:227]: i think having a comment here would be easier than having to scroll to find which col is col25 and why we need to add the trailing whitespace
```suggestion
                    if (col == 25) { \\ col 25 is charr_arr bpchar(5)[], we need to account for the trailing whitespace

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/serde/RecordReaderFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    public void testGetStreamRecordReader() {
+        context.setOutputFormat(OutputFormat.TEXT);
+        assertTrue(factory.getRecordReader(context, true) instanceof StreamRecordReader);
+    }
+
+    @Test
+    public void testGetTestRecordReader() {
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/serde/RecordReaderFactoryTest.java:38]:
```suggestion
    public void testGetTextRecordReader() {

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/serde/StreamRecordReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    public void testReadRecord() throws Exception {
+        StreamRecordReader reader = new StreamRecordReader(new RequestContext());
+        List<OneField> record = reader.readRecord(mockInputStream);
+        assertNotNull(record);
+        assertEquals(1, record.size());
+        assertEquals(DataType.BYTEA.getOID(), record.get(0).type);
+        assertSame(mockInputStream, record.get(0).val);
+        verifyNoInteractions(mockInputStream); // no reading should actually happen
+    }
+}
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/serde/StreamRecordReaderTest.java:34]: nit: missing new line at end of file

File: server/pxf-service/src/test/resources/data/README.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+The files in this directory are auto-generated. Steps to reproduce:
+
+1. Run `GPDataGenerator` from your IDE, it will generate `sample_data.sql` in this directory.
```
Comment (ashuka24) [inline server/pxf-service/src/test/resources/data/README.md:3]: Is there a way to generate the sql without the use of an IDE?

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    }
+
+    /**
+     * Helper function to print the table COPY DDL to the provided PrintStream.
+     * @param out print stream to print output to
+     */
+    private void printCopyDataDML(PrintStream out) {
+        // print out SQL DML statements
+        // use CTAS with ORDER BY to ensure an ordered set of rows in the output file
+        out.println(String.format("COPY (SELECT * FROM sample_data ORDER BY id) TO '%s/sample_data.txt';", dir));
```
Comment (bradfordb-vmware) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java:300]: It looks like this will result in paths that are specific to individual developers machines being included in the generated SQL file and that these will be checked in. Is there any way to either use a path like `/tmp/data` or that references a [psql variable](https://stackoverflow.com/a/18730455) ?

File: server/pxf-service/src/test/resources/data/README.md (hunk 9 above)
Comment (bradfordb-vmware) [inline server/pxf-service/src/test/resources/data/README.md:3]: Can we add a Gradle task for running this?

File: automation/src/main/java/org/greenplum/pxf/automation/components/common/DbSystemObject.java [hunk 11]
```diff
@@ -158,7 +159,23 @@ public void insertData(Table source, Table target) throws Exception {
 				dataStringBuilder.append(",");
 			}
 		}
-		runQuery("INSERT INTO " + target.getName() + " VALUES " + dataStringBuilder);
+		insertData(target, dataStringBuilder.toString());
+	}
+
+	public void insertData(Table target, String data) throws Exception {
```
Comment (bradfordb-vmware) [inline automation/src/main/java/org/greenplum/pxf/automation/components/common/DbSystemObject.java:165]: Since this is public, should we include a JavaDoc, especially for the expected format of the `data` parameter?

File: automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java [hunk 12]
```diff
@@ -112,16 +113,92 @@ private void initPort() {
...
+
+		copyData(sourceName, target, columns,false, null);
+	}
+
+	/**
+	 * Copies data from source table into target table
+	 * @param sourceName name of the source table
+	 * @param target target table
+	 * @param columns columns to select from the source table, if null then all columns will be selected
+	 * @param context extra context to append to the SQL query
```
Comment (bradfordb-vmware) [inline automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java:153]: It looks like context is added before the SQL query not after.

File: automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java [hunk 13]
```diff
@@ -112,16 +113,92 @@ private void initPort() {
...
+	 * @param sourceName name of the source table
+	 * @param target target table
+	 * @param columns columns to select from the source table, if null then all columns will be selected
+	 * @param ignoreFail whether to ignore any failures
+	 * @param context extra context to add to the SQL query
+	 * @throws Exception if the operation fails
+	 */
+	public void copyData(String sourceName, Table target, String[] columns, boolean ignoreFail, String context) throws Exception {
+		String columnList = (columns == null || columns.length == 0) ? "*" : String.join(",", columns);
+		String query = String.format("%sINSERT INTO %s SELECT %s FROM %s",
```
Comment (bradfordb-vmware) [inline automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java:194]: Should we be including the column names in the destination table

```sql
INSERT INTO <table-name>(<column-names>) SELECT <column-names> FROM <source>
```

File: automation/src/main/java/org/greenplum/pxf/automation/structures/tables/pxf/ForeignTable.java [hunk 14]
```diff
@@ -16,7 +16,8 @@ public String constructCreateStmt() {
                 .append(createHeader())
                 .append(createFields())
                 .append(createServer())
-                .append(createOptions());
+                .append(createOptions())
+                .append(distribution());
```
Comment (bradfordb-vmware) [inline automation/src/main/java/org/greenplum/pxf/automation/structures/tables/pxf/ForeignTable.java:20]: Is `DISTRIBUTED BY` supported on foreign tables in GP7? I don't see any mention of it in the [docs](https://docs.vmware.com/en/VMware-Greenplum/7/greenplum-database/ref_guide-sql_commands-CREATE_FOREIGN_TABLE.html).

**Review by bradfordb-vmware — COMMENTED** [review]
Comments from an initial high-level review.

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        // but Univocity tries to allocate the buffer of this size ahead of time, which is very inefficient
+        // parserSettings.setMaxCharsPerColumn(Integer.MAX_VALUE);
+
+        // create the CSV parser with desired settings
+        parser = new CsvParser(parserSettings);
+
+        if (LOG.isDebugEnabled()) {
+            // replace new line characters so that the log message takes only 1 line
+            LOG.debug("Configured CSV Parser : {}", csvFormat.toString()
+                    .replace("\n", " ")
```
[the remark this answers is not in the document]
Comment (denalex) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:82]: What used to get logged:
```
: Configured CSV Parser : CsvFormat:
		Comment character=#
		Field delimiter=,
		Line separator (normalized)=\n
		Line separator sequence=\n
		Quote character="
		Quote escape character="
		Quote escape escape character=null
```
and after substitutions (they use 2 \t chars) it becomes:
```
: Configured CSV Parser : CsvFormat: ||Comment character=# ||Field delimiter=, ||Line separator (normalized)=\n ||Line separator sequence=\n ||Quote character=" ||Quote escape character=" ||Quote escape escape character=null
```

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+                    .replace("\t", "|"));
```
[the remark this answers is not in the document]
Comment (denalex) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:83]: Any suggestions ? same can be said about any other character :) I might just delete it altogether so it will look like (see also comment above):
```
: Configured CSV Parser : CsvFormat: Comment character=# Field delimiter=, Line separator (normalized)=\n Line separator sequence=\n Quote character=" Quote escape character=" Quote escape escape character=null
```

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (bradfordb-vmware) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:82]: I don't think it matters too much but I think if we used `replaceAll("\n\t+", " | ")`, the given example would become:

```text
: Configured CSV Parser : CsvFormat: | Comment character=# | Field delimiter=, | Line separator (normalized)=\n | Line separator sequence=\n | Quote character=" | Quote escape character=" | Quote escape escape character=null
```

**Review by bradfordb-vmware — APPROVED** [review]

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (denalex) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:82]: good idea, did just that.

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 17]
```diff
@@ -73,10 +77,8 @@ public TextRecordReader(RequestContext context, PgUtilities pgUtilities) {
         parser = new CsvParser(parserSettings);
 
         if (LOG.isDebugEnabled()) {
-            // replace new line characters so that the log message takes only 1 line
-            LOG.debug("Configured CSV Parser : {}", csvFormat.toString()
-                    .replace("\n", " ")
-                    .replace("\t", "|"));
+            // replace new line and tab characters so that the log message takes only 1 line
+            LOG.debug("Configured CSV Parser : {}", csvFormat.toString().replaceAll("\n\t+", " | "));
```
Comment (ashuka24) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:81]: previously we were replacing `\n` with ` ` and `\t` with `|`. By changing it to a replace all here, I'm not sure I understand how it does the same mapping of `\n` -> ` ` and `\t` -> `|`

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java [hunk 18]
```diff
@@ -236,7 +233,7 @@ public static void assertDataSet(List<List<OneField>> rows) {
     }
 
     /**
-     * Main method that generates '{project}/server/pxf-service/src/test/resources/data/sample_data.sql' file
+     * Main method that generates '$HOME/workspace/pxf/server/pxf-service/src/test/resources/data/sample_data.sql' file
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java:236]: we set the `dir above` so isn't this just `'${dir}/sample_data.sql`

File: server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java [hunk 19]
```diff
@@ -296,10 +293,16 @@ private void printInsertDataDML(PrintStream out) {
      */
     private void printCopyDataDML(PrintStream out) {
         // print out SQL DML statements
+        // use PSQL variables to dynamically determine the user's home directory and absolute path of the file
+        out.println("\\set data_dir `echo $HOME/workspace/pxf/server/pxf-service/src/test/resources/data/`");
```
Comment (ashuka24) [inline server/pxf-service/src/test/java/org/greenplum/pxf/service/GPDataGenerator.java:297]: can we not use `${dir}` here?

**Review by ashuka24 — APPROVED** [review]

File: automation/src/test/java/org/greenplum/pxf/automation/features/writable/HdfsWritableTextTest.java [hunk 20]
```diff
@@ -558,21 +558,20 @@ public void veryLongRecords() throws Exception {
...
-        writableExTable.setPath(hdfsPath);
-        writableExTable.setFormat("CSV");
         writableExTable.setDistributionFields(new String[]{"key"});
         gpdb.createTableAndVerify(writableExTable);
+        insertData(dataTable, writableExTable, InsertionMethod.INSERT);
 
-        gpdb.insertData(dataTable, writableExTable);
+        if (protocol != ProtocolEnum.HDFS && protocol != ProtocolEnum.FILE) {
+            // for HCFS on Cloud, wait a bit for async write in previous steps to finish
+            sleep(10000);
```
Comment (hpandeycodeit) [inline automation/src/test/java/org/greenplum/pxf/automation/features/writable/HdfsWritableTextTest.java:569]: Is `10s` enough for async?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+import org.greenplum.pxf.api.error.PxfRuntimeException;
+import org.greenplum.pxf.api.io.DataType;
+import org.greenplum.pxf.api.model.GreenplumCSV;
+import org.greenplum.pxf.api.model.RequestContext;
+import org.greenplum.pxf.plugins.hdfs.utilities.PgUtilities;
+
+import java.io.DataInput;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
```
Comment (hpandeycodeit) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:24]: nit: unused import

File: server/pxf-service/build.gradle [hunk 22]
```diff
@@ -39,10 +39,12 @@ dependencies {
     implementation("org.apache.logging.log4j:log4j-spring-boot")
     implementation('org.springframework.boot:spring-boot-starter-actuator')
     implementation('io.micrometer:micrometer-registry-prometheus')
+    implementation("com.univocity:univocity-parsers")
```
Comment (hpandeycodeit) [inline server/pxf-service/build.gradle:42]: This parser hadn't had a release in a while and the repo doesn't look active. What if we face an issue with the parser itself?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java (hunk 17 above)
Comment (denalex) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:81]: The string contained `\n\t\t` after each line, so we're just printing ` | ` instead of this pattern

File: server/pxf-service/build.gradle (hunk 22 above)
Comment (denalex) [reply server/pxf-service/build.gradle:42]: I guess there is not much innovation in CSV parsing. IF for some reason we encounter an unresolvable issue, we can always switch the implementation of the `TextRecordReader` to use another parser library without end-users noticing.

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java (hunk 17 above)
Comment (bradfordb-vmware) [reply server/pxf-service/src/main/java/org/greenplum/pxf/service/serde/TextRecordReader.java:80]: @ashuka24 Take a look at this [comment](https://github.com/greenplum-db/pxf/pull/973#discussion_r1206083897) which has an example of before and after. I think the example should help show what this line is doing now.
